src/box/lua/config/utils/schema.lua

```diff
@@ -1923,6 +1966,63 @@ end
 
 -- }}} Schema preprocessing
 
+-- {{{ Schema validation
+
+-- The unique_items annotation is only applicable to an array of
+-- strings.
+--
+-- We can support other item types in a future, but let's be
+-- conservative on the first step. See validate_unique_items() for
+-- details.
+local function validate_schema_node_unique_items(schema, ctx)
+    if not schema.unique_items then
+        return
+    end
+
+    if schema.type ~= 'array' then
+        walkthrough_error(ctx, '"unique_items" requires an array of ' ..
+            'strings, got a %s', schema.type)
+    end
+
+    if schema.items.type ~= 'string' then
+        walkthrough_error(ctx, '"unique_items" requires an array of ' ..
+            'strings, got an array of %s items', schema.items.type)
+    end
+end
+
+local function validate_schema_impl(schema, ctx)
+    validate_schema_node_unique_items(schema, ctx)
+
+    -- luacheck: ignore 542 empty if branch
+    if is_scalar(schema) then
+        -- Nothing to do.
+    elseif schema.type == 'record' then
+        for field_name, field_def in pairs(schema.fields) do
+            walkthrough_enter(ctx, field_name)
+            validate_schema_impl(field_def, ctx)
+            walkthrough_leave(ctx)
+        end
+    elseif schema.type == 'map' then
+        walkthrough_enter(ctx, '*')
+        validate_schema_impl(schema.key, ctx)
+        validate_schema_impl(schema.value, ctx)
+        walkthrough_leave(ctx)
+    elseif schema.type == 'array' then
+        walkthrough_enter(ctx, '*')
+        validate_schema_impl(schema.items, ctx)
+        walkthrough_leave(ctx)
+    else
+        assert(false)
+    end
```

Comment:
nit: `assert(false)` can be just moved outside the `if..elseif` statement, removing the `else` branch, by adding `return` statement to each `if..elseif`, that way we could remove the luacheck ignore tag, and replace the 'Nothing to do.' comment with a return statement.

Just a suggestion, I'm not sure which approach is clearer.

Reply:
I don't mind any of these two approaches, but I wrote almost all the walkthrough code in this file in the same way, see `*_impl` functions. I hope that if all this code is shaped in the same way, it is easier to understand and to write similar new code.

We possibly can wrap it into some general walkthrough function that invokes callback like `map_enter`, `map_leave`, `record_enter`, `record_leave` and so on. However, I find this code simple enough to don't introduce extra abstraction levels.

Reply:
Ok, I didn't notice other cases) lets leave it as-is.